[dcl.meaning] General wording cleanup

sdkrystian opened the pull request:
The conventions used to specify the meaning of declarators are a little inconsistent, and the sub-clause itself has a lot of single sentence paragraphs that look messy.

For example, [dcl.meaning] has a mix of specifying the "type of the identifier" and "type of the _declarator-id_", [dcl.ptr], [dcl.ref] and [dcl.mptr] determine the "type of the identifier", and [dcl.array] and [dcl.fct] determine the "type of the _declarator-id_". The mentions of "determining the type of the identifier" have been changed to determining the type of the _declarator-id_, as this is a better defined term.

Further, [[dcl.meaning] p3](http://eel.is/c++draft/dcl.meaning#3) correctly acknowledged that `T` can contain an  _attribute-specifier-seq_, but in the very next paragraph it assumes that `T` will just be a _decl-specifier-seq_. [dcl.meaning] p3 is rewritten to account for that, and it additionally specified that the _attribute-specifier-seq_ appertains to the declarator (this is also specified in [[dcl.pre] p2](http://eel.is/c++draft/dcl.pre#2.sentence-5), perhaps that mention could be removed, as it makes more sense to have it where the meaning of declarators is determined).

More stylistically, [dcl.meaning] uses `codeblock` when specifying a declaration of the for `T D`, which needlessly uses whitespace, and is inconsistent with the other sub-clauses. These were changed to use `\tcode`.

Lastly, [the last sentence of [dcl.meaning] p6](http://eel.is/c++draft/dcl.meaning#6.sentence-2) is completely redundant, and should be just turned into a note.

sdkrystian committed:
Clean up [dcl.meaning]

jensmaurer commented on source/declarations.tex:
```diff
@@ -2563,13 +2565,23 @@
+the type of the
+\grammarterm{declarator-id}
+in
+\tcode{D}
```
I think  "in D" can go here; there is only one declarator-id we talk about in this context.

jensmaurer replied:
Ah, no, that's the pattern we use elsewhere, too.

jensmaurer commented on source/declarations.tex:
```diff
@@ -2563,13 +2565,23 @@
+\tcode{D}
+is ``\tcode{T}''.
+The optional
+\grammarterm{attribute-specifier-seq}
```
Since this seems to be new text, please follow the "semantic linebreaks" conventions. In particular, don't start a new line when a \xxx LaTeX thing occurs. (These are historic accidents from when we did the standard in troff.)

sdkrystian replied:
Should I change this for all the modified lines? Or just for new text.

jensmaurer replied:
As a first step, the usual green blob

then the type of the
\grammarterm{declarator-id}
in

should be on one line, and also the corresponding blob above (maybe that needs two lines).
Also
\tcode{T}
\tcode{D}
should be on one line for the new next.

Please don't wholesale-change the unchanged stuff; that should be part of a different change.

jensmaurer reviewed (approved):
Looks fine (and is an improvement). It would be good if the line breaks wouldn't follow a troff tradition.

sdkrystian committed:
Fix style

zygoloid commented on source/declarations.tex:
```diff
@@ -2521,34 +2520,24 @@
-\tcode{T}
-determines the type
-\tcode{T}.
+First, the type \tcode{T} is determined.
```
We have two different things we're calling `T` here: a *decl-specifier-seq* and a type. This rewrite makes it less clear that that's happening by removing the "the *decl-specifier-seq* `T` determines the type `T`" part. I think we should take this a bit further. How about replacing the previous paragraph ("Thus[...]") with:

> Thus, a declaration of a particular identifier has the form *attribute-specifier-seq*[opt] *decl-specifier-seq* *declarator*. The optional *attribute-specifier-seq* appertains to the declared entity. Following is a recursive procedure for determining the type of the declared entity.

Then we only have one `T`, and it's a type, not a *decl-specifier-seq*. So we can start this second paragraph with something like:

> First, a type `T` is determined from the *decl-specifier-seq*. [Example: ...

sdkrystian replied:
Perhaps we want to talk about the type of the *declarator-id* rather than the declared entity? Declarations introduce names (which denote entities), after all.

zygoloid commented on source/declarations.tex:
```diff
@@ -2562,37 +2551,33 @@
+the type of the \grammarterm{declarator-id}
+in \tcode{D} is ``\tcode{T}''.
+The optional \grammarterm{attribute-specifier-seq}
+appertains to the entity declared by \tcode{D}.
```
We don't need the "In a declaration" part here. How about reducing this to:

> A *declarator* of the form *declarator-id* *attribute-specifier-seq*[opt] determines the type of its declared entity to be `T`. The optional *attribute-specifier-seq* appertains to the declared entity.

sdkrystian replied:
Its true that its not quite needed, however it makes it consistent with the rest of [dcl.meaning]

sdkrystian replied:
See my comments on maintaining `T` through subclauses

zygoloid commented on source/declarations.tex:
```diff
@@ -2562,37 +2551,33 @@
 Parentheses do not alter the type of the embedded
 \grammarterm{declarator-id},
 but they can alter the binding of complex declarators.
+\end{note}
```
> A *declarator* of the form `(` *D* `)` determines the type of its declared entity to be the same as the type determined by the *declarator* *D*. [Note: ...]

jensmaurer replied:
The problem is that a declarator (alone) does not determine any type at all. Example: `(*p)` is a declarator, but doesn't actually say what the type is: "pointer to whatever" is not a type. The rephrasing here departs from the careful constructions involved derived-declarator-type-list.

zygoloid commented on source/declarations.tex:
```diff
@@ -2607,12 +2592,16 @@
 \tcode{T}
 \tcode{D1}
 is ``\placeholder{derived-declarator-type-list}
 \tcode{T}'',
```
These, again, are two completely different `T`s. =(

sdkrystian replied:
I think that there is no issue here - the point is to use the same decl-specifier-seq in invented declaration `T D1`.

zygoloid commented on source/declarations.tex:
```diff
@@ -2607,12 +2592,16 @@
+in
 \tcode{D}
 is ``\placeholder{derived-declarator-type-list} \grammarterm{cv-qualifier-seq} pointer to
 \tcode{T}''.
```
> A *declarator* of the form `*` *attribute-specifier-seq*[opt] *cv-qualifier-seq* *D* determines the type of its declared entity to be the type of the entity declared by the declaration *U* *D*, where *U* is a *type-specifier* that denotes the type ''*cv-qualifier-seq* pointer to `T`''.

zygoloid commented on source/declarations.tex:
```diff
@@ -2713,12 +2702,16 @@
+in
 \tcode{D}
 is ``\placeholder{derived-declarator-type-list} reference to
 \tcode{T}''.
```
> A *declarator* of the form `&` *attribute-specifier-seq*[opt] *D* or `&&` *attribute-specifier-seq*[opt] *D* determines the type of its declared entity to be the type of the entity declared by the declaration *U* *D*, where *U* is a *type-specifier* that denotes the type ''reference to `T`''.

(And so on for later subclauses.)

sdkrystian replied:
@zygoloid The removal of _derived-declarator-type-list_ is definitely a good idea, however, moving the form of the declarator to be inline does degrade readability, especially for longer declarator forms (i.e. functions). Perhaps we keep the form of the declarators indented as is done currently, and apply the other changes to the paragraphs?

sdkrystian replied:
Also, maintaining `T` throughout subclauses is a little awkward, and could lead to confusion (typically, when "variables" are established, we carry them no further than then end of a paragraph, let alone a subclause)

zygoloid reviewed (changes requested):
I think we can clean this up quite a bit more. Suggestion follows.

jensmaurer commented:
While I very much appreciate a general cleanup in this area, I'm mildly favoring a core issue for this, or at least CWG review of the editorial change. @zygoloid , are you reasonably sure we're not destroying `int (*p)()` vs. `int *p()` with the reformulation? (The reformulation looks reasonable to me, but this is not the first time in history that CWG was bitten by this wording. Maybe that's an argument for refomulating.) Note that we've lived with the status quo since C++98, so I'm slightly uneasy about a last-minute fix for C++20.
Please don't read this as a veto, just as a "github diff format isn't actually the right tool for this kind of change".

sdkrystian committed:
Apply suggested changes

sdkrystian commented:
@zygoloid I applied the changes you suggested, but rather than referring to "the declared entity", I changed it to refer to the _declarator-id_ itself.

sdkrystian commented:
@jensmaurer I walked through a few complex declarators manually following the new wording and it all checks out.

tkoeppe commented:
@sdkrystian: could you please rebase this (maybe after the mailing goes out), and let me know, and then I'll ask CWG to review this. The cleanup looks laudable, but I'd like some committee oversight on it.

sdkrystian commented:
@jensmaurer @tkoeppe I apologize for going dark for the last few years... I will be going through my open PRs addressing comments/rebasing over the coming weeks.
